src/core/public/osd_bootstrap.ts

```diff
@@ -31,13 +31,41 @@
 import { i18n } from '@osd/i18n';
 import { CoreSystem } from './core_system';
 import { ApmSystem } from './apm_system';
+import { getAndUpdateLocaleInUrl } from './locale_helper';
 
 /** @internal */
 export async function __osdBootstrap__() {
   const injectedMetadata = JSON.parse(
     document.querySelector('osd-injected-metadata')!.getAttribute('data')!
   );
 
+  // Extract the locale from the URL if present
+  // This allows for dynamic locale setting via URL parameters
+  const urlLocale = getAndUpdateLocaleInUrl(window.location.href);
+
+  if (urlLocale) {
+    // If a locale is specified in the URL, update the i18n settings
+    // This enables dynamic language switching
+    // Note: This works in conjunction with server-side changes:
+    // 1. The server registers all available translation files at startup
+    // 2. A server route handles requests for specific locale translations
+
+    // Set the locale in the i18n core
+    // This will affect all subsequent i18n.translate() calls
+    i18n.setLocale(urlLocale);
+
+    // Modify the translationsUrl to include the new locale
+    // This ensures that the correct translation file is requested from the server
+    // The replace function changes the locale in the URL, e.g.,
+    // from '/translations/en.json' to '/translations/zh-CN.json'
+    injectedMetadata.i18n.translationsUrl = injectedMetadata.i18n.translationsUrl.replace(
+      /\/([^/]+)\.json$/,
+      `/${urlLocale}.json`
```

Comment:
if user sets an invalid locale through URL, is it the same behavior as setting it through config/opensearch_dashboards.yml and OSD UI won't break?

Reply:
It will break just as you set an invalid one in config/opensearch_dashboards.yml. Both will use load in i18n 
```
export async function load(translationsUrl: string) {
  // Once this package is integrated into core OpenSearch Dashboards we should switch to an abstraction
  // around `fetch` provided by the platform, e.g. `kfetch`.
  const response = await fetch(translationsUrl, {
    credentials: 'same-origin',
  });

  if (response.status >= 300) {
    throw new Error(`Translations request failed with status code: ${response.status}`);
  }

  init(await response.json());
}
```
this fetch will call locale route in src/legacy/ui/ui_render/ui_render_mixin.js
```
server.route({
    path: '/translations/{locale}.json',
    method: 'GET',
    config: { auth: false },
    handler: async (request, h) => {
      const { locale } = request.params;
      const normalizedLocale = locale.toLowerCase();
      const registeredLocales = i18nLoader.getRegisteredLocales().map((l) => l.toLowerCase());

      // Function to get or create cached translations
      const getCachedTranslations = async (localeKey, getTranslationsFn) => {
        if (!translationsCache[localeKey]) {
          const translations = await getTranslationsFn();
          translationsCache[localeKey] = {
            translations: JSON.stringify(translations),
            hash: createHash('sha1').update(JSON.stringify(translations)).digest('hex'),
          };
        }
        return translationsCache[localeKey];
      };

      let cachedTranslations;

      if (normalizedLocale === defaultLocale.toLowerCase()) {
        // Default locale
        cachedTranslations = await getCachedTranslations(defaultLocale, () =>
          i18n.getTranslation()
        );
      } else if (registeredLocales.includes(normalizedLocale)) {
        // Other registered locales
        cachedTranslations = await getCachedTranslations(normalizedLocale, () =>
          i18nLoader.getTranslationsByLocale(locale)
        );
      } else {
        // Locale not found
        throw Boom.notFound(`Unknown locale: ${locale}`);
      }

      return h
        .response(cachedTranslations.translations)
        .header('cache-control', 'must-revalidate')
        .header('content-type', 'application/json')
        .etag(cachedTranslations.hash);
    },
  });
```
for config change it will go to clause `normalizedLocale === defaultLocale.toLowerCase()` and for dynamic it will go to `else if (registeredLocales.includes(normalizedLocale))` because defaultLocale is the current i18n's locale. So if we have a dynamic locale, we will check if it has a translation file otherwise will throw `throw Boom.notFound(`Unknown locale: ${locale}`);`, which is same as if you pass an invalid locale to config.

Reply:
got it, thanks

Reply:
sorry. my comment is not correct. I think there is a bug in the previous behavior: once we defined in the config this `i18n.getLocale() !== locale.toLowerCase()` is always true so we never call ` throw Boom.notFound(`Unknown locale: ${locale}`);`
```
handler(request, h) {
      // OpenSearch Dashboards server loads translations only for a single locale
      // that is specified in `i18n.locale` config value.
      const { locale } = request.params;
      if (i18n.getLocale() !== locale.toLowerCase()) {
        throw Boom.notFound(`Unknown locale: ${locale}`);
      }

      // Stringifying thousands of labels and calculating hash on the resulting
      // string can be expensive so it makes sense to do it once and cache.
      if (translationsCache.translations == null) {
        translationsCache.translations = JSON.stringify(i18n.getTranslation());
        translationsCache.hash = createHash('sha1')
          .update(translationsCache.translations)
          .digest('hex');
      }

      return h
        .response(translationsCache.translations)
        .header('cache-control', 'must-revalidate')
        .header('content-type', 'application/json')
        .etag(translationsCache.hash);
    },
```

I am not sure if this a proper behavior because user has no idea that they pass the wrong locale. If we want to stick to this behavior we could change to 
```
else {
        // Locale not found
        cachedTranslations = await getCachedTranslations('en', () =>
          i18nLoader.getTranslationsByLocale('en')
        );
      }
```